pvlib/irradiance.py

```diff
@@ -1182,7 +1182,98 @@ def perez(surface_tilt, surface_azimuth, dhi, dni, dni_extra,
         return sky_diffuse
 
 
-def disc(ghi, zenith, datetime_or_doy, pressure=101325):
+def clearness_index(ghi, solar_zenith, extra_radiation, min_cos_zenith=0.065):
+    """
+    Calculate the clearness index.
+
+    The clearness index is the ratio of global to extraterrestrial
+    irradiance on a horizontal plane.
+
+    Parameters
+    ----------
+    ghi : numeric
+        Global horizontal irradiance in W/m^2.
+
+    solar_zenith : numeric
+        True (not refraction-corrected) solar zenith angle in decimal
+        degrees.
+
+    extra_radiation : numeric
+        Irradiance incident at the top of the atmosphere
+
+    min_cos_zenith : numeric, default 0.065
+        Minimum value of cos(zenith) to allow when calculating global
+        clearness index `kt`. Equivalent to zenith = 86.273 degrees.
+
+    Returns
+    -------
+    kt : numeric
+        Clearness index
+
+    Reference
+    ---------
+    [1] Maxwell, E. L., "A Quasi-Physical Model for Converting Hourly
+    Global Horizontal to Direct Normal Insolation", Technical
+    Report No. SERI/TR-215-3087, Golden, CO: Solar Energy Research
+    Institute, 1987.
+    """
+    cos_zenith = tools.cosd(solar_zenith)
+    I0h = extra_radiation * np.maximum(cos_zenith, min_cos_zenith)
+    # consider adding
+    # with np.errstate(invalid='ignore', divide='ignore'):
+    # to kt calculation, but perhaps it's good to allow these
+    # warnings to the users that override min_cos_zenith
+    kt = ghi / I0h
+    kt = np.maximum(kt, 0)
+    # Limit copied from the kt prime limit in dirint, which was justified
+    # with reference to SRRL code. consider replacing with 1 or nan
+    kt = np.minimum(kt, 0.82)
```

Comment:
I've looked into the source of 0.82 as the upper bound.  It does not appear in the references, which don't specify an upper bound.  The value 0.82 comes from a Fortran implementation of the DIRINT model at NREL, and appears to have been added as a modification to the original algorithm.

I recommend we keep the upper bound, but make it a kwarg (like `min_cos_zenith`) with default of 0.82, and point at NREL SRRL's Fortran code for DIRINT as the source (although the file appears no longer available through their website.)

An upper bound makes sense, but with the caution that the value will depend on the time averaging.  0.82 is probably an hourly value.  For a minute, clearness index > 1 is certainly possible if the GHI is enhanced by cloud effects.

Reply:
Thanks for the good comment. We should definitely make a kwarg. I'd recommend a maximum value of 2 for a few reasons. My general feeling is that the defaults should be as permissive as possible and the user has some responsibility to QA/QC their data before putting it into this function. Also, subhourly data seems to be common these days.

Reply:
2 is OK with me.  G. Yordanov has a few conference papers looking at the magnitude, duration and frequency of occurrence of over-irradiance, down to sub-second responses from a photodiode sensor. From memory, a maximum value of 2 would be a generous limit.  2 would likely also prevent the overflow warning reported in #540

Reply:
I will add the kwarg to ``clearness_index`` and ``clearness_index_zenith_independent`` with default values of 2. Should it also be a kwarg in ``disc``, ``dirint``, ``dirindex`` (like ``min_cos_zenith`` and ``max_zenith``)?

Reply:
They do call ``clearness_index`` but that doesn't mean that we have to expose all of its options. It's not clear to me if ``disc``, ``dirint``, ``dirindex`` should enforce a different limit.

Reply:
Right, they do with this pull request.  I think that `disc` and `dirint` should enforce a limit on `kt`.  `dirindex` should The challenge here is that the limit is up to us.  The only precedent we have is old SRRL Fortran code for `dirint` which is where 0.82 comes from.  I'm fairly sure that value is appropriate for hourly data but not for shorter time steps.  I agree with you that kt<2 is generally applicable and is better than 0.82.

I'd say let's go ahead and expose the kwarg.

Reply:
Updating the unit tests has me thinking that we should enforce a limit of 0.82 or 1 or we should return NaN. 

For ``disc``, the input conditions that yield ``kt > 1.25`` will produce ``dni = 0``. This because ``Kn`` becomes negative and we set negative values to 0.

For ``dirint``, the input conditions that yield ``kt_prime > 1`` will produce ``dni = NaN``.

Here's the line in ``dirint`` that currently limits ``kt_prime`` to 0.82:

https://github.com/pvlib/pvlib-python/blob/master/pvlib/irradiance.py#L1358

These lines assign bins for ``kt_prime`` values [0, 1]:

https://github.com/pvlib/pvlib-python/blob/master/pvlib/irradiance.py#L1381-L1387

These lines will ensure that ``kt_prime`` values > 1 will become NaN:

https://github.com/pvlib/pvlib-python/blob/master/pvlib/irradiance.py#L1425-L1428

One could argue that returning NaN is the right thing to do because the authors did not anticipate the model being applied for clearness index > 1. Or one could argue that we should limit the clearness index to 0.82 or 1 and move on.

Reply:
I think you are suggesting that the call to `clearness_index` in, e.g., `disc`, passes a value `max_kt`=1 (or otherwise) that is appropriate for `disc`. If this is correct, then I'm OK.  If you are suggesting that the default value for `max_kt` is set to prevent wonky output from `disc` then I'm not OK.

The appropriate value for each model is less clear.  The only source we have is the SRRL code for `dirint` which is an uncommented addition to the original program.  Since I am confident that none of these models meant to predict DNI from GHI during over-irradiance conditions, an upper bound of 0.82, or 1, seems appropriate to me.

On the other hand, for `clearness_index` itself, I like your suggestion of `max_kt`=2, for physical reasons.

I don't think `disc` etc. should return `NaN` unless its inputs are meaningless, or in the case of `dirint` are outside the model's domain of bins.

Reply:
Sorry for being unclear. We agree (I think): within ``disc`` and ``dirint``, we should call ``clearness_index`` and ``clearness_index_zenith_independent`` with ``max_clearness_index=1``. The functions retain the default max value of 2. I'll update the PR this morning.